components/UI/Datalist.js

```diff
@@ -0,0 +1,114 @@
+import React, { useState, useRef } from 'react';
+
+import { ServiceTable } from './ServiceTable';
+
+/**
+ * An input component that functions as a datalist and can be controlled with arrow keys and mouse clicks.
+ *
+ * @component
+ * @param {string[]} options - An array of Objects for the datalist.
+ * @param {function} onChange - function to execute when input changes, receives string as first parameter.
+ * @param {function} onSelect - function to execute when a result from the datalist was selected
+ * @param {Array} keysToDisplay - Array of strings of key values to be displayed as results
+ * @returns {React.JSX.Element} The Datalist component.
+ */
+
+function Datalist({ options, onChange, onSelect, keysToDisplay }) {
+    const [value, setValue] = useState('');
+    const [isOpen, setIsOpen] = useState(false);
+    const [isLoading, setIsLoading] = useState(false);
+    const [selectedIndex, setSelectedIndex] = useState(-1);
+    const inputRef = useRef(null);
+
+    const handleChange = async (e) => {
+        setIsLoading(true);
+        setValue(e.target.value);
+        onSelect(null);
+        await onChange(e.target.value);
+        if (e.target.value !== '') setIsOpen(true);
+        else {
+            // if the input is empty we close the datalist
+            setIsOpen(false);
+        }
+        setIsLoading(false);
+    };
+
+    const handleKeyDown = (e) => {
+        // Handle keyboard navigation
+        if (e.key === 'ArrowDown') {
+            e.preventDefault();
+            if (!isOpen) {
+                setIsOpen(true);
+            } else {
+                const newIndex = Math.min(selectedIndex + 1, options.length - 1);
```

Comment:
No need to define a `const` really here, or is there?
If you move this right into the line below, then there's one less variable for our JavaScript memory to keep track of.

Same goes for the lines below.